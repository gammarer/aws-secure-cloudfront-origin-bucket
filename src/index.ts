import { SecureBucket, SecureBucketEncryption } from '@gammarer/aws-secure-bucket';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

export interface SecureCloudFrontOriginBucketProps {
  readonly bucketName?: string;
  readonly cloudFrontOriginAccessIdentityS3CanonicalUserId: string;
}

export class SecureCloudFrontOriginBucket extends SecureBucket {

  constructor(scope: Construct, id: string, props: SecureCloudFrontOriginBucketProps) {
    super(scope, id, {
      bucketName: props.bucketName,
      encryption: SecureBucketEncryption.S3_MANAGED, // Noticeï¼‰Only S3 Managed
      versioned: false,
    });

    // ðŸ‘‡Allow CloudFront access
    this.addToResourcePolicy(new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ['s3:GetObject'],
      principals: [
        new iam.CanonicalUserPrincipal(
          props.cloudFrontOriginAccessIdentityS3CanonicalUserId,
        ),
      ],
      resources: [`${this.bucketArn}/*`],
    }));
  }
}
