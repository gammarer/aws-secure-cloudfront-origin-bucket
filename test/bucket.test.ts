import { App, Stack } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { SecureCloudFrontOriginBucket } from '../src';


it('Is Bucket', () => {
  const app = new App();
  const stack = new Stack(app, 'TestingStack', {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
  });

  const bucket = new SecureCloudFrontOriginBucket(stack, 'SecureCloudFrontOriginBucket', {
    bucketName: 'origin-bucket',
    cloudFrontOriginAccessIdentityS3CanonicalUserId: '',
  });
  expect(bucket).toBeInstanceOf(s3.Bucket);
  const assert = Template.fromStack(stack);
  assert.hasResource('AWS::S3::Bucket', {});
});

it('Has managed encrypted Bucket', () => {
  const app = new App();
  const stack = new Stack(app, 'TestingStack');

  new SecureCloudFrontOriginBucket(stack, 'testing', {
    bucketName: 'origin-bucket',
    cloudFrontOriginAccessIdentityS3CanonicalUserId: '',
  });
  const template = Template.fromStack(stack);

  template.hasResourceProperties('AWS::S3::Bucket', {
    BucketEncryption: Match.objectEquals({
      ServerSideEncryptionConfiguration: [
        {
          ServerSideEncryptionByDefault: {
            SSEAlgorithm: 'AES256',
          },
        },
      ],
    }),
  });
});

it('Has BlockPublicAccess to BLOCK_ALL', () => {
  const app = new App();
  const stack = new Stack(app, 'TestingStack');

  new SecureCloudFrontOriginBucket(stack, 'testing', {
    bucketName: 'origin-bucket',
    cloudFrontOriginAccessIdentityS3CanonicalUserId: '',
  });

  const template = Template.fromStack(stack);

  template.hasResourceProperties('AWS::S3::Bucket', {
    PublicAccessBlockConfiguration: Match.objectEquals({
      BlockPublicAcls: true,
      BlockPublicPolicy: true,
      IgnorePublicAcls: true,
      RestrictPublicBuckets: true,
    }),
  });
});

it('Does not allow for unencrypted uploads', () => {
  const app = new App();
  const stack = new Stack(app, 'TestingStack');

  new SecureCloudFrontOriginBucket(stack, 'TestingBucket', {
    bucketName: 'origin-bucket',
    cloudFrontOriginAccessIdentityS3CanonicalUserId: '',
  });

  const template = Template.fromStack(stack);

  template.hasResourceProperties('AWS::S3::BucketPolicy', {
    Bucket: {
      Ref: Match.stringLikeRegexp('TestingBucket'),
    },
    PolicyDocument: {
      Version: '2012-10-17',
      Statement: Match.arrayWith([
        Match.objectLike({
          Action: 's3:*',
          Condition: {
            Bool: {
              'aws:SecureTransport': 'false',
            },
          },
          Effect: 'Deny',
          Principal: {
            AWS: '*',
          },
        }),
      ]),
    },
  });

});

it('Allow from CloudFront Origin Access Identity', () => {
  const app = new App();
  const stack = new Stack(app, 'TestingStack');

  new SecureCloudFrontOriginBucket(stack, 'TestingBucket', {
    bucketName: 'origin-bucket',
    cloudFrontOriginAccessIdentityS3CanonicalUserId: 'CloudFront Origin Access Identity XXXXXXXXXXXXXX',
  });

  const template = Template.fromStack(stack);

  template.hasResourceProperties('AWS::S3::BucketPolicy', {
    Bucket: {
      Ref: Match.stringLikeRegexp('TestingBucket'),
    },
    PolicyDocument: {
      Version: '2012-10-17',
      Statement: Match.arrayWith([
        Match.objectLike({
          Action: 's3:GetObject',
          Effect: 'Allow',
          Principal: {
            CanonicalUser: 'CloudFront Origin Access Identity XXXXXXXXXXXXXX',
          },
        }),
      ]),
    },
  });
});
