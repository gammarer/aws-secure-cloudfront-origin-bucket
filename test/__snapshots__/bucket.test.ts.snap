// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecureCloudFrontOriginBucket Testing Should match snapshot: bucket 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "TestingBucket7BD4A29B": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AccessControl": "Private",
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": "origin-bucket",
        "OwnershipControls": Object {
          "Rules": Array [
            Object {
              "ObjectOwnership": "BucketOwnerEnforced",
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestingBucketPolicy566DA0C0": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "TestingBucket7BD4A29B",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TestingBucket7BD4A29B",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "TestingBucket7BD4A29B",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": "CloudFront Origin Access Identity XXXXXXXXXXXXXX",
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "TestingBucket7BD4A29B",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
